airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/load/test/util/destination_process/DockerizedDestination.kt

```diff
@@ -4,85 +4,261 @@
+        // Certainly nothing in the bulk CDK's test suites is reading back
+        // anything in this directory.
+        val localRoot = Files.createTempDirectory(testDir, "output")
+        // This directory will contain the actual inputs to the connector (config+catalog),
+        // and is also mounted as a volume.
+        val jobRoot = Files.createDirectories(workspaceRoot.resolve("job"))
+
+        // Extract the string "destination-foo" from "gcr.io/airbyte/destination-foo:1.2.3".
+        // The old code had a ton of extra logic here, along with a max string
+        // length (docker container names must be <128 chars) - none of that
+        // seems necessary here.
+        // And platform doesn't even follow that convention anymore, now that
+        // we have monopods. (the pod has a name like replication-job-18386126-attempt-4,
+        // and the destination container is just called "destination")
+        val shortImageName = imageTag.substringAfterLast("/").substringBefore(":")
+        val containerName = "$shortImageName-$command-$randomSuffix"
+        logger.info { "Creating docker container $containerName" }
+
+        val cmd: MutableList<String> =
+            Lists.newArrayList(
+                "docker",
+                "run",
+                "--rm",
+                "--init",
+                "-i",
+                "-w",
+                "/data/job",
+                "--log-driver",
+                "none",
+                "--name",
+                containerName,
+                "--network",
+                "host",
+                "-v",
+                String.format("%s:%s", workspaceRoot, "/data"),
+                "-v",
+                String.format("%s:%s", localRoot, "/local"),
+                "-e",
+                "DEPLOYMENT_MODE=$testDeploymentMode",
+                // Yes, we hardcode the job ID to 0.
+                // Also yes, this is available in the configured catalog
+                // via the syncId property.
+                // Also also yes, we're relying on this env var >.>
+                "-e",
+                "WORKER_JOB_ID=0",
+                imageTag,
+                command,
+            )
+
+        fun addInput(paramName: String, fileContents: Any) {
+            Files.write(
+                jobRoot.resolve("destination_$paramName.json"),
+                Jsons.writeValueAsBytes(fileContents),
+            )
+            cmd.add("--$paramName")
+            cmd.add("destination_$paramName.json")
+        }
+        config?.let { addInput("config", it) }
+        catalog?.let { addInput("catalog", it) }
+
+        logger.info { "Executing command: ${cmd.joinToString(" ")}" }
+        process = ProcessBuilder(cmd).start()
+        // Annoyingly, the process's stdin is called "outputStream"
+        destinationStdin = BufferedWriter(OutputStreamWriter(process.outputStream))
+    }
+
     override fun run() {
-        TODO("launch a docker container")
+        runBlocking {
```

Comment:
this is maybe bad? Should this function just be a suspend fun?

Reply:
It's bad because you're calling it from inside of other suspend funs:

```        
    fun runSync(
...
    ): List<AirbyteMessage> {
...
        return runBlocking {
            val destinationCompletion = async { destination.run() }
            messages.forEach { destination.sendMessage(it.asProtocolMessage()) }
```

[It blocks the current **thread** until completion](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/run-blocking.html).

If you're planning to keep this pattern, then I'd make `run` a suspend fun.

Reply:
`runBlocking` should only be used as a top-level bridge between non-suspend fun's and suspend funs.

Reply:
done in https://github.com/airbytehq/airbyte/pull/46363/commits/00ff4fe92e1d2d4c26b5d3cb3a936fbd02d9b69c